ratatui-widgets/src/table.rs

```diff
@@ -2276,4 +2315,157 @@ mod tests {
+            Rect::new(0, 0, 5, render_height),
+            &mut buf,
+            &mut state,
+        );
+
+        assert_eq!(expected, buf);
+    }
+
+    /// If there isn't enough room for the selected item and the requested padding the table can
+    /// jump up and down every frame if something isn't done about it. This code tests to make
+    /// sure that isn't currently happening
+    #[test]
+    fn padding_flicker() {
+        let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 5));
+        let mut state = TableState::default();
+
+        *state.offset_mut() = 2;
+        state.select(Some(4));
+
+        let rows = (0..8).map(|i| Row::new([i.to_string()]));
+        let list = Table::new(rows, [Constraint::Length(2)])
+            .scroll_padding(3)
+            .highlight_symbol(">> ");
+
+        StatefulWidget::render(&list, buffer.area, &mut buffer, &mut state);
+
+        let offset_after_render = state.offset();
+
+        StatefulWidget::render(&list, buffer.area, &mut buffer, &mut state);
+
+        // Offset after rendering twice should remain the same as after once
+        assert_eq!(offset_after_render, state.offset());
+    }
+
+    #[test]
+    fn padding_inconsistent_item_sizes() {
+        let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 3));
+        let mut state = TableState::default().with_offset(0).with_selected(Some(3));
+
+        let items = [
+            Row::new(["Item 0"]),
+            Row::new(["Item 1"]),
+            Row::new(["Item 2"]),
+            Row::new(["Item 3"]),
+            Row::new(["Item 4\nTest\nTest"]),
+            Row::new(["Item 5"]),
+        ];
+        let list = Table::new(items, [Constraint::Length(7)])
+            .scroll_padding(1)
+            .highlight_symbol(">> ");
+
+        StatefulWidget::render(list, buffer.area, &mut buffer, &mut state);
+
+        #[rustfmt::skip]
+        let expected = [
+            "   Item 2 ",
+            ">> Item 3 ",
+            "   Item 4 ",
+        ];
+        assert_eq!(buffer, Buffer::with_lines(expected));
+    }
+
+    // Tests to make sure when it's pushing back the first visible index value that it doesnt
+    // include an item that's too large
+    #[test]
+    fn padding_offset_pushback_break() {
+        let mut buffer = Buffer::empty(Rect::new(0, 0, 10, 4));
+        let mut state = TableState::default();
+
+        *state.offset_mut() = 1;
+        state.select(Some(2));
+
+        let items = [
+            Row::new(["Item 0\nTest\nTest"]),
+            Row::new(["Item 1"]),
+            Row::new(["Item 2"]),
+            Row::new(["Item 3"]),
+        ];
+        let table = Table::new(items, [Constraint::Length(10)])
+            .scroll_padding(2)
+            .highlight_symbol(">> ");
+
+        StatefulWidget::render(table, buffer.area, &mut buffer, &mut state);
+        #[rustfmt::skip]
+        assert_eq!(
+            buffer,
+            Buffer::with_lines([
+                "   Item 0 ",
```

Comment:
This doesn't seem reasonable - Item 0 is truncated here... Am I missing something?